### PR TITLE
[WIP] Move selector matching to apimachinery

### PR DESCRIPTION
**Pull request by karlkfi**
#### What?

- Add external label and field selector implementations to
	k8s.io/api/*
- Add internal label and field selector implementations to
	k8s.io/kubernetes/pkg/apis/*
- Expand unit testing of label selector conversion functions to
	also test the new field selector functions in
	k8s.io/kubernetes/pkg/api. This ensures all the external versioned
	selectable fields are converted to selectable internal fields.
- Add runtime.Scheme.AddSelectorFunc and Matches for registering
	SelectorFuncs for each GVK and using them for matching objects,
	implementing label and field selectors generically.
- Add runtime.Selectors and runtime.SelectionPredicate to help
	implement label and fields selection matching.
- Modify storage.SelectionPredicate to wrap runtime.SelectionPredicate.
- Add unit tests to ensure all internal types with custom field
	selectors only have field selectors that can be converted to valid
	internal field selectors.
- Update all fake clients to implement label and field selectors on
	both List and Watch.
- Add testing.ObjectWatchReaction to build WatchReactionFuncs that
	wrap the ObjectTracker, similar to testing.ObjectReaction for
	ReactionFuncs.
- Update clientgen and generated clients to use ObjectWatchReaction
- Fix multiple bugs related to not being able to select fields that
	should be selectable because they weren't specified in the field
	conversions.
- Move SchemeBuilder.Register calls from register.go to the files
	that define the methods being registered. This makes it easier
	to write custom conversions and defaults by copying from the
	previous version and modifying.
- Fix a bug in the reflecter watch cache that was calling
	watch.ResultChan() on every iteration of the for loop. This was
	causing deadlocks any time you use a watch.Interface implementation
	that returns a new channel on each call to ResultChan().

#### Why?

My initial goal was to just update FakeDynamicClient to support label and field selectors, for use in testing a downstream change in cli-utils's status watcher: https://github.com/kubernetes-sigs/cli-utils/pull/639

However, after some investigation, it seems that the field selector matcher code was buried deep in the registry code and inaccessible to the FakeDynamicClient, or anybody else. The registry injects the Matcher/SelectionPredicate and GetAttrs functions into the resource-specific REST storage code, but it seems excessive to need to build the whole registry to simulate selectors, and they're hidden behind private fields anyway.

With these changes, field and label selection can be replicated by fake client implementations using the exact same code as the apiserver, which avoids needing to use integration/e2e tests with a full apiserver to test controllers and clients that use selectors.

#### Dependencies

- https://github.com/kubernetes/kubernetes/pull/125204
- https://github.com/kubernetes/kubernetes/pull/125107

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

It's possible may want to note the fixed field selection bugs and changes to the fake client, which may require downstream users (like controller test) to update in order to specify fields being selected that the fake client List & Watch previously ignored. Let me know.

/sig api-machinery
/kind cleanup

### CODE REVIEW

**Comment by karlkfi** [comment]
/cc @sttts @smarterclayton 

Yall are owners on a bunch of the packages being touched here.

**Comment by alexzielenski** [comment]
I think this moves us in the right direction. The selectable fields logic is not something always directly mapped to JSON paths, and itd be great to test them directly.

Some things worth exploring we discussed offline a bit:

1. Definining a new interface for types to implement rather than registering selector functions (ala zz_generated_lifecycle) We have precedent for both styles (defaults uses function registration). Curious about reasons to go for one style over the other here i.e. implement `ToSelectableFields()` on each type. 
2. Implementing for versioned types rather than unversioned. One thing to consider if we end up implementing these for versioned types without conversion is how to keep all implementations in sync

cc @liggitt or @deads2k for comment on high-level approach

**Comment by karlkfi** [comment]
> The selectable fields logic is not something always directly mapped to JSON paths, and itd be great to test them directly.

I actually found and fixed a few bugs when adding tests for the field selectors and field converters. There were several cases where the field conversion funcs (used by `Scheme.ConvertFieldLabel`) would error because they didn't recognize a field that was supposed to be selectable (according to the `SelectionPredicate` registered with the `registry.Store.PredicateFunc` and `generic.StoreOptions.AttrFunc`  ).

> Implementing for versioned types rather than unversioned. One thing to consider if we end up implementing these for versioned types without conversion is how to keep all implementations in sync

I have added versioned and internal selector to `pkg/apis`. I also added and upgraded the unit tests in `pkg/api/testing` to loop through all the fields from the `[versioned].ToSelectableFields`, run them through `Scheme.ConvertFieldLabel` to convert them to unversioned fields, and then compare them to the output of the `[internal].ToSelectableFields`. This is how I found some existing bugs with the field conversion fields not matching the field selector fields. The old tests were only checking that the field conversion was 1:1 with itself except for hard-coded exceptions in the tests. And half the types and versions with custom conversions weren't being tested before.

> Definining a new interface for types to implement rather than registering selector functions (ala zz_generated_lifecycle) We have precedent for both styles (defaults uses function registration). Curious about reasons to go for one style over the other here i.e. implement ToSelectableFields() on each type.

This could work. Putting it on the type itself might be simpler than having to register separate SelectorFuncs with the scheme. You don't need to import the internal types the same way you do to build the conversion code.

**Comment by karlkfi** [comment]
Figured out why a lot of my tests were timing out...

There's a subtle bug in cache.Reflector that was calling watch.ResultChan() on every iteration of the for loop, instead of saving it to a local variable outside the loop. 

This was causing deadlocks if you use a watch.Interface impl that returns a new channel on every call to ResultChan().

This explains why all the other internal impls save the result channel to the watch struct and return the same one each time. But it means you can't use the same watch to fan-out and receive all the events on multiple handlers. And it means it starts a goroutine prematurely when the watch is constructed, instead of waiting until ResultChan is called.

**Comment by karlkfi** [comment]
I also found a bug that when you filter a watch by label or field selector, if an object is updated/patched to remove one of the selected labels or fields, the watch filter converts the Modified event into a Delete event.

This has the added problem that Delete events give you the previous object state, instead of the new object state. So you can't even check the object to see if a required label or field was removed. You have to make a whole Get call to check if the object was actually deleted or not.

Oddly, I found a warning (7 years old) on the `watch.Filter` about this, but the current filteredWatch doesn't actually perform the translation from Modified to Deleted, and nothing in k/k is actually using it. So the comment must be legacy and the translation behavior must have been preserved and refactored to elsewhere in the code at some point.

**Comment by liggitt** [comment]
> There's a subtle bug in cache.Reflector that was calling watch.ResultChan() on every iteration of the for loop, instead of saving it to a local variable outside the loop.
> 
> This was causing deadlocks if you use a watch.Interface impl that returns a new channel on every call to ResultChan().

That seems like a bug in that ResultChan() impl to me... calling ResultChan() twice should be ok

**Comment by liggitt** [comment]
> I also found a bug that when you filter a watch by label or field selector, if an object is updated/patched to remove one of the selected labels or fields, the watch filter converts the Modified event into a Delete event.

That's not a bug, the object *was* "deleted" from the set of filtered objects because it no longer matched. The delete event is how the object gets removed from the informer cache, etc.

> This has the added problem that Delete events give you the previous object state, instead of the new object state. So you can't even check the object to see if a required label or field was removed. You have to make a whole Get call to check if the object was actually deleted or not.

If you want to distinguish between API object deletion and modifications that make an object no longer match a given selector, you can't delegate filtering to the server... you have to list / watch unfiltered and check for whether the object has the label or field you are interested in.

**Comment by karlkfi** [comment]
> That seems like a bug in that ResultChan() impl to me... calling ResultChan() twice should be ok

Calling ResultChan() twice isn't what deadlocks. But if you implement ResultChan() such that it makes a new channel and a new go routine to send events to that channel, and the caller only ever reads one event from any given result channel, and the result channel isn't buffered, then the events back up on the 2nd event, which is never read from the first result channel that was returned.

Like I said, it's a bit subtle. This is actually very similar to problems I've seen when blocking on "done channels" in a for|select loop. As long as the Done() method always returns the same channel, you can call it in the select case directly, but if if returns a new channel each time it's called, then you have to call `doneCh := Done()` before the for loop and use the local `doneCh` in the select case. 

Reflector is doing this:
```
	for {
		select {
		case <-stopCh:
			return errorStopRequested
		case err := <-errc:
			return err
		case event, ok := <-w.ResultChan():
			if !ok {
				break loop
			}
			...
		}
	}
```

Instead of this:
```
	resultCh := w.ResultChan()
	for {
		select {
		case <-stopCh:
			return errorStopRequested
		case err := <-errc:
			return err
		case event, ok := <-resultCh:
			if !ok {
				break loop
			}
			...
		}
	}
```

The `watch.Interface.ResultChan()` comments don't document what the expected behavior is when called multiple times, so it's relatively easy to write an impl that doesn't conform to existing usage.

**Comment by karlkfi** [comment]
> That's not a bug, the object was "deleted" from the set of filtered objects because it no longer matched. The delete event is how the object gets removed from the informer cache, etc.

This may be intended behavior, and it's beyond the scope of this PR to change, but this behavior is definitely inefficient, especially when paired with an informer cache. The caller is forced to choose between a potentially much larger informer cache, containing objects they don't care about with overlapping resource type, or making more API calls to confirm deletion before deciding how to handle a delete event.

**Comment by karlkfi** [comment]
After moving the selection.go impls from internal to public API packages, I had to update some unit tests to populate `spec.nodeName` on Pods to make the selectors work in the fake client, using the real Scheme. So that shows that the field selectors are working in the fake client.

File: cmd/kubeadm/app/util/config/cluster_test.go [hunk 1]
```diff
@@ -364,6 +364,7 @@ func TestGetAPIEndpointWithBackoff(t *testing.T) {
 			name:     "valid ipv4 endpoint in pod annotation",
 			nodeName: nodeName,
 			staticPod: &testresources.FakeStaticPod{
+				NodeName:  nodeName,
```
Comment (neolit123) [inline cmd/kubeadm/app/util/config/cluster_test.go:367]: why are these NodeName changes in the cmd/kubeadm file needed?

Comment (karlkfi) [reply cmd/kubeadm/app/util/config/cluster_test.go:367]: The fake client being used now respects field selectors, which are being specified by the controller. So the node names need to be specified to be returned by the list/watch.

**Comment by karlkfi** [comment]
Now that tests are passing and I think this works as intended, I've extracted out some of the watcher fixes to another PR: https://github.com/kubernetes/kubernetes/pull/125107

**Pull request closed without merging**